projects/plugins/jetpack/modules/memberships/class-jetpack-memberships.php

```diff
@@ -329,16 +329,31 @@ public function render_button( $attributes, $content = null, $block = null ) {
 	public function get_subscription_url( $plan_id ) {
 		global $wp;
 
-		return add_query_arg(
-			array(
-				'blog'     => esc_attr( self::get_blog_id() ),
-				'plan'     => esc_attr( $plan_id ),
-				'lang'     => esc_attr( get_locale() ),
-				'pid'      => esc_attr( get_the_ID() ), // Needed for analytics purposes.
-				'redirect' => esc_attr( rawurlencode( home_url( $wp->request ) ) ), // Needed for redirect back in case of redirect-based flow.
-			),
-			'https://subscribe.wordpress.com/memberships/'
+		$query_args = array(
+			'blog'     => esc_attr( self::get_blog_id() ),
+			'plan'     => esc_attr( $plan_id ),
+			'lang'     => esc_attr( get_locale() ),
+			'pid'      => esc_attr( get_the_ID() ), // Needed for analytics purposes.
+			'redirect' => esc_attr( rawurlencode( home_url( $wp->request ) ) ), // Needed for redirect back in case of redirect-based flow.
 		);
+
+		if ( ! defined( 'IS_WPCOM' ) || ! IS_WPCOM ) {
+			return add_query_arg( $query_args, 'https://subscribe.wordpress.com/memberships/' );
+		}
+
+		require_lib( 'memberships' );
+		$plan = Memberships_Product::get_from_post( self::get_blog_id(), $plan_id );
```

Comment:
Calling Memberships_Product::get_from_post() on the front-end (on the visitor facing part of the site) might be risky from a performance PoV? 🤔

Reply:
>The proposed solution is to fetch the membership data when building the URL while rendering each payment button. I've checked that the amount of queries added by this solution scales, we are only performing one extra query and it's indexed so it shouldn't cause any performance issues, the code is only executed for pages that have an actual payment button.

That's true, but `get_from_post` does a switch_blog function call which AFAIK is a heavy operation. This would aggravate things when we have more than one button on a page.

Reply:
In the context of these changes the switch_to_blog function always switches to the same blog, which skips all the heavy parts.